dhis-2/dhis-services/dhis-service-data-exchange/src/main/java/org/hisp/dhis/dataexchange/aggregate/AggregateDataExchangeService.java

```diff
@@ -87,34 +94,73 @@ public AggregateDataExchangeService( AnalyticsService analyticsService,
         this.encryptor = encryptor;
     }
 
+    @Transactional( readOnly = true )
+    public AggregateDataExchange getById( String uid )
+    {
+        return aggregateDataExchangeStore.loadByUid( uid );
+    }
+
     /**
      * Runs the analytics data exchange with the given identifier.
      *
      * @param uid the {@link AggregateDataExchange} identifier.
+     * @param progress {@link JobProgress} to track progress when running in a
+     *        job context
      * @return an {@link ImportSummaries}.
      */
-    public ImportSummaries exchangeData( String uid )
+    @Transactional
+    public ImportSummaries exchangeData( String uid, JobProgress progress )
     {
         AggregateDataExchange exchange = aggregateDataExchangeStore.loadByUid( uid );
 
-        return exchangeData( exchange );
+        return exchangeData( exchange, progress );
     }
 
     /**
      * Runs the given analytics data exchange.
      *
-     * @param uid the {@link AggregateDataExchange}.
+     * @param exchange the {@link AggregateDataExchange}.
+     * @param progress {@link JobProgress} to track progress when running in a
+     *        job context
      * @return an {@link ImportSummaries}.
      */
-    public ImportSummaries exchangeData( AggregateDataExchange exchange )
+    @Transactional
+    public ImportSummaries exchangeData( AggregateDataExchange exchange, JobProgress progress )
     {
         ImportSummaries summaries = new ImportSummaries();
 
-        exchange.getSource().getRequests()
-            .forEach( request -> summaries.addImportSummary( exchangeData( exchange, request ) ) );
-
-        log.info( "Aggregate data exchange completed: '{}', type: '{}'",
-            exchange.getUid(), exchange.getTarget().getType() );
+        Target target = exchange.getTarget();
+        Api api = target.getApi();
+        progress.startingStage( format( "exchange aggregate data %s to %s target %s", exchange.getName(),
+            target.getType().name().toLowerCase(), api == null ? "(local)" : api.getUrl() ),
+            FailurePolicy.SKIP_ITEM );
+        Function<SourceRequest, ImportSummary> runExchange = request -> {
+            try
+            {
+                return exchangeData( exchange, request );
+            }
+            catch ( Exception ex )
+            {
+                ImportSummary summary = new ImportSummary();
+                summary.setStatus( ImportStatus.ERROR );
+                summary.setDescription( ex.getMessage() );
+                return summary;
+            }
+        };
+
+        progress.runStage( exchange.getSource().getRequests().stream(),
```

Comment:
Looks good. Perhaps reduce the mental load of reading this part a little bit by splitting out to methods or named functions (just a suggestion)?

Reply:
Yes :D I extracted quite a number of methods. There was way to much going on.